### PR TITLE
refactor: bring back config attribute

### PR DESCRIPTION
**Pull request by thesayyn**
TLDR;

- move authentication code into it's own file
- remove auth_locator repository rule 
- un-deprecate config attribute
- make oci_pull find the config file, and allow it to be overridden via a label attribute. (this will allow users to pass a file from workspace)

This is my third take on config file location detection. This change will cause oci_pull and oci_alias to perform auth file location determination. 

Now instead of performing the config file location determination only once through config_file_locator repo rule, we do it everytime for every oci_pull rule. 

#### This rules out the chance that we will get a stale result for config path and try to read the nonexsistent file. 

This has happened before in the past because Aspect workflows was reusing repository rule caches across machines. and that's the reason why we started using file path instead of just copying file to repository rule or symlinking it)

####  The original motivation for performing auth file location once was to reduce noise generated by that logic, eg printing "auth file couldn't be found". 

I brought back this problem again but now it's less annoying as we print that message only when pull fails.

### CODE REVIEW

**Comment by alexeagle** [comment]
Is this meant as an alternative to #493 ?

**Comment by mrmeku** [comment]
@thesayyn Don't forget https://github.com/bazel-contrib/rules_oci/blob/main/oci/extensions.bzl :)
Thanks so much for the hard work!

**Comment by thesayyn** [comment]
> Is this meant as an alternative to #493 ?

Yes, this is principled way of doing that instead of playing with semantics of DOCKER_CONFIG.

**Comment by thesayyn** [comment]
@mrmeku does this solve your relative path issue?

**Comment by mrmeku** [comment]
Yep!

File: oci/private/authn.bzl [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    "docker.elastic.co": {
+        "realm": "docker-auth.elastic.co/auth",
+        "scope": "repository:{repository}:pull",
+        "service": "token-service",
+    },
+}
+
+def _strip_host(url):
+    # TODO: a principled way of doing this
+    return url.replace("http://", "").replace("https://", "").replace("/v1/", "")
```
Comment (alexeagle) [inline oci/private/authn.bzl:55]: nit: should it use string.stripprefix ?

**Review by mrmeku — APPROVED** [review]

File: oci/private/authn.bzl [hunk 2]
```diff
@@ -0,0 +1,249 @@
+"repository rule that locates the .docker/config.json or containers/auth.json file."
+
+load("@aspect_bazel_lib//lib:base64.bzl", "base64")
+load("@aspect_bazel_lib//lib:repo_utils.bzl", "repo_utils")
+load(":util.bzl", "util")
+
+
+# Unfortunately bazel downloader doesn't let us sniff the WWW-Authenticate header, therefore we need to
+# keep a map of known registries that require us to acquire a temporary token for authentication.
+_WWW_AUTH = {
```
[the remark this answers is not in the document]
Comment (thesayyn) [reply oci/private/authn.bzl:10]: Yes, that is on my list.

**Review by alexeagle — APPROVED** [review]
Thanks for helping with review @mrmeku
